python/mujoco/rollout.py

```diff
@@ -74,132 +302,38 @@ def rollout(
-      control=control,
-      state=state,
-      sensordata=sensordata,
-  )
-  if isinstance(model, list) and nroll == 1:
-    nroll = len(model)
-
-  if isinstance(model, list) and len(model) != nroll:
-    raise ValueError(
-        f'nroll inferred as {nroll} but model is length {len(model)}'
-    )
-  elif not isinstance(model, list):
-    model = [model]  # Use a length 1 list to simplify code below
-
-  # infer nstep, check for incompatibilities
-  nstep = _infer_dimension(
-      1, nstep or 1, control=control, state=state, sensordata=sensordata
-  )
-
-  # get nstate/ncontrol/nv/nsensordata
-  # check that they are equal across models
-  nstate = mujoco.mj_stateSize(
-      model[0], mujoco.mjtState.mjSTATE_FULLPHYSICS.value
-  )
-  ncontrol = mujoco.mj_stateSize(model[0], control_spec)
-  nv = model[0].nv
-  nsensordata = model[0].nsensordata
-  for m in model[1:]:
-    if (
-        nstate
-        != mujoco.mj_stateSize(m, mujoco.mjtState.mjSTATE_FULLPHYSICS.value)
-        or ncontrol != mujoco.mj_stateSize(m, control_spec)
-        or nv != m.nv
-        or nsensordata != m.nsensordata
-    ):
-      raise ValueError('models are not compatible')
-
-  # check trailing dimensions
-  _check_trailing_dimension(nstate, initial_state=initial_state, state=state)
-  _check_trailing_dimension(ncontrol, control=control)
-  _check_trailing_dimension(nv, initial_warmstart=initial_warmstart)
-  _check_trailing_dimension(nsensordata, sensordata=sensordata)
-
-  # tile input arrays/lists if required (singleton expansion)
-  model = model * nroll if len(model) == 1 else model
-  initial_state = _tile_if_required(initial_state, nroll)
-  initial_warmstart = _tile_if_required(initial_warmstart, nroll)
-  control = _tile_if_required(control, nroll, nstep)
-
-  # allocate output if not provided
-  if state is None:
-    state = np.empty((nroll, nstep, nstate))
-  if sensordata is None:
-    sensordata = np.empty((nroll, nstep, nsensordata))
-
-  # call rollout
-  _rollout.rollout(
-      model,
-      data,
-      nstep,
-      control_spec,
-      initial_state,
-      initial_warmstart,
-      control,
-      state,
-      sensordata,
-  )
+  if not isinstance(data, list):
+    data = [data]  # Use a length 1 list to simplify code below
+  nthread = len(data) if len(data) > 1 else 0
+
+  # Use a persistent thread pool if requested
+  if persistent_pool:
+    global persistent_rollout
+    # Create or restart persistent threadpool
+    if persistent_rollout is None or persistent_rollout.nthread != nthread:
+      persistent_rollout = Rollout(nthread=nthread)
+    rollout = persistent_rollout
+  else:
+    rollout = Rollout(nthread=nthread)
+
+  ret = rollout.rollout(
+    model,
+    data,
+    initial_state,
+    control,
+    control_spec=control_spec,
+    skip_checks=skip_checks,
+    nstep=nstep,
+    initial_warmstart=initial_warmstart,
+    state=state,
+    sensordata=sensordata,
+    chunk_size=chunk_size)
+
+  if not persistent_pool:
```

Comment:
Can you put this in a `try ... finally` block, so `close()` is called even if an exception is thrown?